manta-accounting/src/wallet/signer/functions.rs

```diff
@@ -286,6 +290,72 @@ where
     }
 }
 
+/// Sums all the values with the same [`Asset`] id in `assets`.
+fn sum_values<C>(assets: &mut Vec<Asset<C>>)
+where
+    C: Configuration,
+    C::AssetValue: Add<Output = C::AssetValue>,
+{
+    let mut result = Vec::new();
+
+    for asset in assets.iter() {
+        if let Some(entry) = result.iter_mut().find(|(id, _)| *id == asset.id) {
+            entry.1 += asset.value.clone();
+        } else {
+            result.push((asset.id.clone(), asset.value.clone()));
+        }
+    }
+
+    *assets = result
+        .into_iter()
+        .map(|(id, value)| Asset::<C>::new(id, value))
+        .collect();
+}
+
+/// First it runs [`sum_values`] in both `deposit` and `withdraw`. Then, for each [`Asset`] id
+/// which happens in both `deposit` and `withdraw`:
+/// 1) computes the difference `diff = asset_value(deposit) - asset_value(withdraw)`
+/// 2) If `diff > 0`, it replaces the corresponding entry in `deposit` with `diff` and deletes
+/// the entry in `withdraw`.
+/// 3) If `diff < 0`, it replaces the corresponding entry in `withdraw` with `-diff` and deletes
+/// the entry in `deposit`.
+/// 4) If `diff = 0`, it deletes the entry in `deposit` and `withdraw`.
+fn normalize_assets<C>(deposit: &mut Vec<Asset<C>>, withdraw: &mut Vec<Asset<C>>)
+where
+    C: Configuration,
+    C::AssetValue: Add<Output = C::AssetValue>,
+    C::AssetValue: Sub<Output = C::AssetValue>,
+{
+    sum_values::<C>(deposit);
+    sum_values::<C>(withdraw);
+    let mut i = 0;
+    while i < deposit.len() {
+        let deposit_asset = deposit[i].clone();
+        if let Some(withdraw_index) = withdraw
+            .iter()
+            .position(|asset| asset.id == deposit_asset.id)
+        {
+            let withdraw_asset = &mut withdraw[withdraw_index];
+            if deposit_asset.value > withdraw_asset.value {
+                let diff = deposit_asset.value.clone() - withdraw_asset.value.clone();
```

Comment:
do we need use `saturating_sub` or `checked_sub` so that if operation overflow, it throw error?